### ISSUE TITLE
Unable to remove attributes from element whose name has accents from different culture

### ISSUE THREAD

**Issue by zivkan** [AngleSharp/AngleSharp#419]
The following code snippet will result in an infinite loop

```
HtmlParser parser = new HtmlParser();
var document = parser.Parse("<TAG MÍ=\"\" />");
var element = (IElement)document.Body.FirstChild;
while (element.Attributes.Length > 0)
{
    element.RemoveAttribute(element.Attributes[0].Name);
}
```

I didn't have time to investigate deeply, but I noticed in Element.cs there's two calls to `ToLower` and two to `ToLowerInvariant`. I don't know any Spanish, but I think the Í character is capital, so my guess is that some parts of the code is converting it to lowercase but some parts are not. What particularly surprises me is that no exception is being thrown, as I see what appears to be some error checking in NamedNodeMap.

Finally, thanks for a really great library. It's been really helpful.

**Comment by FlorianRappl** [comment]
Thanks for the report - I'll try to find these nasty `ToLower` calls and make them behave correctly!

Will be fixed for the upcoming release.

**Comment by mausworks** [comment]
@FlorianRappl what about comparing strings using something like:

```
attributeName.Equals(other, StringComparison.OrdinalIgnoreCase)
```

_Or some other `StringComparison` kind, not sure which kind is best suited for this._

`ToLower`-calls, usually never behave correctly. 😢

**Comment by FlorianRappl** [comment]
No offense, but I guess you never checked out the code of AngleSharp (or the code that is addressed with this issue). The `ToLower` here is **not** used for comparison.

**Comment by mausworks** [comment]
None taken, @FlorianRappl. I did not mean to insult the code base in any way, so I beg pardon. I _have_ checked out the code, I love this project. But I've not read every single line of code.

I was simply replying to your statement about nasty `ToLower`-calls, and making them behave correctly.

[`ToLower` is being used](https://github.com/AngleSharp/AngleSharp/blob/fe3eb2702c2da4cc9cb75c60fdc0fb8a68e13a18/src/AngleSharp/Dom/Element.cs#L497) though, and as such, _is_ being used for comparison... just further down the road [in `RemoveNamedItemOrDefault`](https://github.com/AngleSharp/AngleSharp/blob/157a10ef0b9615cabb3309bd5f4ba28bb575ade8/src/AngleSharp/Dom/Collections/NamedNodeMap.cs#L75).

With that said... It doesn't seem to be the issue as `"MÍ".ToLower()` &rarr; `"mí"`. It normalizes just fine.

**Comment by FlorianRappl** [comment]
Here `ToLower` is used to accord to the specification. Attributes that are within the HTML namespace need to be transformed to lower-case. Therefore, if querying against HTML attributes the generic comparison algorithm is used (case-sensitive), with an input parameter unknown of the namespace. Thus, if the incoming namespace is HTML as well, the transformation takes care of comparing the right things.

So what I referred to is that we do not use `ToLower` to fake some case-insensitive comparison, but rather to follow the specification. The spec. dictates that HTML attributes are only stored lower case. It also dictates that (only) HTML attributes are compared in a case-insensitive way. So the same transformation that is applied for storing, needs to be applied to the input to compare the right things.

If we would only care about HTML then I would totally feel fine about only using the comparison (however, serialization would then have to use `ToLower` again - quite ugly), but some elements may be SVG, MathML, or even elements from some other namespace. AngleSharp carefully follows the spec. here, meaning the transformation was a necessary evil.

I'll see what I can do (the invariant path is the one I want to take). Maybe there is a more efficient way to solve this (e.g., boolean flag to determine if case-insensitive comparison should be used).

**Comment by mausworks** [comment]
Thank you for your wonderful breakdown @FlorianRappl 😄 

You could create something like a `IAttributeNameTransformer` and then implement `DefaultAttributeNameTransformer` (which is basically no-op) and `HtmlAttributeTransformer` (which might just initially call `.ToLower()`).

**Something like this.**

```
public interface IAttributeNameTransformer
{
    string Transform(string attributeName);
}
```

```
public interface HtmlAttributeNameTransformer
{
    public string Transform(string attributeName) => attributeName.ToLower();
}
```

```
public interface DefaultAttributeNameTransformer
{
    public string Transform(string attributeName) => attributeName;
}
```

This might feel a bit verbose though. On the plus-side &ndash; it can be passed in the constructor of `Element` or even be a property of `Document` (leaky). It's nice and testable, and you don't even need to check the NS.

**Example:**

```
public void RemoveAttribute(String name)
{
    _attributes.RemoveNamedItemOrDefault(_attributeNameTransformer.Transform(name));
}
```

You can tell me when I'm getting out of line here...

**Comment by FlorianRappl** [comment]
I think the way you describe would add way too much overhead. Also why don't I need to check for the namespace?

Where should the interface live? It has to be applied to every attribute then. But it also needs a way in the element. Both, the element and the attribute may have one (but not necessarily the same) namespace. So you could have an SVG element with an HTML attribute. Similarly, you could search for an SVG attribute within an HTML element (containing any kind of attributes). (So we now have 3 possible locations / resolutions for this: in the attribute (required), in the element (default), in the search (explicit / optional).)

Generally, AngleSharp tries to avoid unnecessary allocations when possible. Just introducing another field (i.e., another allocation) for this purpose is too much. There are DOMs with >= 10k elements. Therefore we easily have 10^4 new fields, i.e., 80 kB only in references (assuming all point to the same instance).

(Plus the described way would make any selectors on attributes slower; you have one additional virtual call that originated from resolving a reference.)

**Comment by mausworks** [comment]
> But it also needs a way in the element. Both, the element and the attribute may have one (but not necessarily the same) namespace. So you could have an SVG element with an HTML attribute. Similarly, you could search for an SVG attribute within an HTML element (containing any kind of attributes).

This breaks it. You are absolutely correct. Thank you.

> Generally, AngleSharp tries to avoid unnecessary allocations when possible. Just introducing another field (i.e., another allocation) for this purpose is too much. There are DOMs with >= 10k elements. Therefore we easily have 10^4 new fields, i.e., 80 kB only in references (assuming all point to the same instance).

The idea is only efficient if it can live as a member of the Owner document.

> (Plus the described way would make any selectors on attributes slower; you have one additional virtual call that originated from resolving a reference.)

Interesting. Method calls are usually measured in nanoseconds though. Still, point taken.

**Comment by FlorianRappl** [comment]
Yes, method calls are fast, but C# is an indirect language. Hence you'll pay for every reference by a potential cache miss. Of course, the CLR does a pretty good job on hiding / minimizing this, but in hot paths (and the attribute one is certainly such a hot path, e.g., for `QuerySelector`) this may add up. I am not sure and the architectural elegance may in the end just outweigh the performance savings (for most code this is certainly the case).

That's why I always run the performance benchmarks before/after such a change (i.e., I am not sure about the impact; all I did was explaining my fears / another arguments against that change - which would indeed by more elegant).

Long story short: Thanks for your input; even though it potentially is not applicable in the described scenario it is certainly helpful in others.

**Comment by FlorianRappl** [comment]
I now know the source of the error. It is tricky ...

I tried running this in the browser

``` html
<!DOCTYPE html>
<html>

<body>
<TAG MÍ="" />
<script>
var element = document.body.firstElementChild;
var name = element.attributes[0].name;
alert(name);
alert(element.attributes.length);
element.removeAttribute(name);
alert(element.attributes.length);
</script>
</body>

</html>
```

and apparently the result is `mÍ` / `1` / `1` with **Chromium** (e.g., Chrome, Opera). However, **IE** and **Firefox** seem to handle this well (we get ... / `0`).

HTML defines what is considered uppercase for the tokenizer (on a char-by-char basis), but does not really refer to this range for the DOM model. It is now clear, however, that the same rules need to be applied (should have been obvious, but I seemed to miss that connection).

In my naive way of thinking I used the `ToLower` (or `ToLowerInvariant`) methods from .NET here, but I need to provide my own extension method following the W3C specification for the definition of "what is an upper char".

I also filed a bug for Chromium (see https://bugs.chromium.org/p/chromium/issues/detail?id=651946). I don't know if this will be fixed there.

**Comment by FlorianRappl** [comment]
Fixed, see PR #421.

**Comment by zivkan** [comment]
Thank you very much!

Would you also consider API changes?

I lost considerable time trying to find the root-cause of this issue, which could have been avoided in several ways. For example:
- Have an API to clear the NamedNodeMap, like every container in the .NET standard library has (although this would have solved my problem only, not the general cause).
- Change `Element.cs` to use `RemoveNamedItem` instead of `RemoveNamedItemOrDefault` (maybe this was done to mimic javascript's API, I don't know. But I'm not a javascript developer and had expectations that this API would work similarly to other .NET APIs that throw exceptions when the key is not found).
- Add an overload, or a new method with an easily discoverable name, so developers can opt-in to .NET-like behaviour instead of javascript-like behaviour.

I now realise that I could have called `element.Attributes.RemoveNamedItem` instead of `element.RemoveAttribute`. But, since `RemoveAttribute`'s XMLdoc doesn't say anything about silently ignoring errors, and no overloads exist, I had no way to guess this would be the behaviour without reading the source code.

**Comment by FlorianRappl** [comment]
AngleSharp uses primarily the official DOM API (naming may be adjusted to .NET conventions, but the behavior is 100% the same - which is by design).

Extension methods exist for convenience, but they always build upon the official DOM API.

**Comment by zivkan** [comment]
While the documentation I could find on w3c.org doesn't say anything about not throwing exceptions, [this page on developer.mozzila.org](https://developer.mozilla.org/en-US/docs/Web/API/Element/removeAttribute) does explicitly state it. Can I suggest that you add it into the xmldoc so that it appears in intellisense when devs like me use it?

I agree that sticking to the official DOM API is the best choice. But in practise we're using it in the .NET ecosystem, which has its own conventions (you already said naming may be adjusted), and not everyone comes from a web-development background so won't know the DOM API. In my case, manipulating HTML is only a very small part of the overall system, so it's not productive for me to learn the details of the DOM API. As a native English speaker, it wasn't intuitive for me to test capitalised non-English attribute names, so I didn't detect this problem until we ran on production data (fortunately in a testing environment) and my infinite loop had a much higher cost than if an exception was thrown or my program generated incorrect output.

Anyway, I really do appreciate you making the library available and with a licence that I'm allowed to use. Thank you very much. Have a great weekend.

**Comment by FlorianRappl** [comment]
I understand that this frustrates you, but when you use a library that was made to embrace the web you should also embrace the web.

The reasoning was simple: I know many (everyone?) who uses this lib will have a .NET background, but a large fraction won't have a background. Those who have a web background will feel familiar (small learning curve). Those who don't have a web background will appreciate that the learning curve they'll have to follow anyway (new lib) will come with the benefit of learning the web APIs in the same process. Thus, its a 2 in 1 win.

I don't know any .NET container which throws an exception on removing an element that is not there. Neither dictionary nor a list does this. Granted - you receive the status back (in form of a boolean), but that's it.

The bug you ran into was indeed a nasty one - and even though a boolean return (which I may add - it is indeed helpful, common in the .NET world, and I see no point in sticking with no return value when one could be helpful) may have been helpful it potentially would have still resulted in that infinite loop. There you pass on the name of an existing attribute to remove an attribute by its name. I mean - what could possible go wrong here?

I'll also add an extension method to clean the attributes. But those two things are "improvements" and therefore not part of this issue.

Thanks again for the report and the suggestions - they are indeed very valuable!

**Comment by zivkan** [comment]
My apologies. I really thought that List<> and Dictionary<> threw exceptions when asked to remove an item that doesn't exist. But I just tested and they don't. They only do for lookups. Well, dictionary and hashmap lookups. Lists use indexes, so it's more obvious for consumers of the API to check index values.  So, please excuse my previous messages. It turns out they were out of frustration, even if I thought I was being more constructive than I turned out to be.

In any case, I've now learned a new useful test case for anything I write in the future, which I'm unlikely to forget any time soon 😄

**Comment by FlorianRappl** [comment]
No problem - I am happy we both gained some insights in the process and (even better) the issue you reported will potentially improve Chromium used by billions of people.
